s3fs/core.py

```diff
@@ -1349,13 +1366,37 @@ def _fetch(self, start, end):
                 return
             if not self.fill_cache and start > self.end:
                 self.start, self.end = None, None
-                return self._fetch(start, end)
+                return self._not_file_backed_cache_fetch(start, end)
             new = _fetch_range(self.s3.s3, self.bucket, self.key, version_id,
                                self.end, end + self.blocksize,
                                req_kw=self.s3.req_kw)
             self.end = end + self.blocksize
             self.cache = self.cache + new
 
+    def _file_backed_cache_fetch(self, start, end):
+        if self.s3.version_aware:
+            version_id = self.version_id
+        else:
+            version_id = None
+        if end > self.end:
+            end = self.end
+        start_block = start // self.blocksize
+        end_block = end // self.blocksize
+        for i in range(start_block, end_block + 1):
```

Comment:
This loop may be wasteful if requesting contiguous blocks, often the overhead of establishing the connection it comparable to the download time; instead of doing a fetch for each block, they should be combined where possible.

Reply:
totally agreed. i just want to show how it works it really should check and merge small requests like this to be as large as possible.